pkg/storegateway/bucket.go

```diff
@@ -615,102 +620,269 @@ func (s *BucketStore) Series(req *storepb.SeriesRequest, srv storepb.Store_Serie
-			}
-			series.Labels = mimirpb.FromLabelsToLabelAdapters(lset)
+// sendStreamingSeriesLabelsHintsStats sends the labels of the streaming series.
+// Since hints and stats need to be sent before the "end of stream" streaming series message,
+// this function also sends the hints and the stats.
+func (s *BucketStore) sendStreamingSeriesLabelsHintsStats(
+	req *storepb.SeriesRequest,
+	srv storepb.Store_SeriesServer,
+	stats *safeQueryStats,
+	seriesSet storepb.SeriesSet,
+	resHints *hintspb.SeriesResponseHints,
+) error {
+	// TODO: should we pool the seriesBuffer/seriesBatch?
+	seriesBuffer := make([]*storepb.StreamingSeries, req.StreamingChunksBatchSize)
+	for i := range seriesBuffer {
+		seriesBuffer[i] = &storepb.StreamingSeries{}
+	}
+	seriesBatch := &storepb.StreamSeriesBatch{
+		Series: seriesBuffer[:0],
+	}
+	// TODO: can we send this in parallel while we start fetching the chunks below?
+	for seriesSet.Next() {
+		var lset labels.Labels
+		// IMPORTANT: do not retain the memory returned by seriesSet.At() beyond this loop cycle
+		// because the subsequent call to seriesSet.Next() may release it.
+		// TODO: check if it is safe to hold the lset.
+		lset, _ = seriesSet.At()
+
+		// We are re-using the slice for every batch this way.
+		seriesBatch.Series = seriesBatch.Series[:len(seriesBatch.Series)+1]
+		seriesBatch.Series[len(seriesBatch.Series)-1].Labels = mimirpb.FromLabelsToLabelAdapters(lset)
 
-			// Encode the message. We encode it ourselves into a PreparedMsg in order to measure
-			// the time it takes.
-			encodeBegin := time.Now()
+		// TODO: Add relevant trace spans and timers.
+
+		if len(seriesBatch.Series) == int(req.StreamingChunksBatchSize) {
 			msg := &grpc.PreparedMsg{}
-			if err = msg.Encode(srv, storepb.NewSeriesResponse(&series)); err != nil {
-				err = status.Error(codes.Internal, errors.Wrap(err, "encode series response").Error())
-				return
+			if err := msg.Encode(srv, storepb.NewStreamSeriesResponse(seriesBatch)); err != nil {
+				return status.Error(codes.Internal, errors.Wrap(err, "encode streaming series response").Error())
 			}
-			encodeDuration += time.Since(encodeBegin)
 
 			// Send the message.
-			sendBegin := time.Now()
-			if err = srv.SendMsg(msg); err != nil {
-				err = status.Error(codes.Unknown, errors.Wrap(err, "send series response").Error())
-				return
+			if err := srv.SendMsg(msg); err != nil {
+				return status.Error(codes.Unknown, errors.Wrap(err, "send streaming series response").Error())
 			}
-			sendDuration += time.Since(sendBegin)
-		}
-		if seriesSet.Err() != nil {
-			err = errors.Wrap(seriesSet.Err(), "expand series set")
-			return
+
+			seriesBatch.Series = seriesBatch.Series[:0]
 		}
+	}
+	if seriesSet.Err() != nil {
+		return errors.Wrap(seriesSet.Err(), "expand series set")
+	}
+
+	// We need to send hints and stats before sending the chunks.
+	// Also, these need to be sent before we send IsEndOfSeriesStream=true.
```

Comment:
What's the reason for this?

Reply:
In case of streaming, queriers fetch the chunk lazily, i.e. when the PromQL is executing. At that point, we should be done with receiving all the series and other stats and hints because we will be out of that function that does that.

Reply:
this sounds like something that the queries should be prepared to handle IMO. If it's too difficult to change, then can you please document it in the proto files and in the querier code (and keep the comment here)?